File: leet_LIST_1669_merge.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def mergeInBetween(self, list1: ListNode, a: int, b: int, list2: ListNode) -> ListNode:
        pA, pB, head = list1, list1, list1
        p2end = list2
        while a > 1 and pA:
            a -= 1
            b -= 1
            pA = pA.next
            pB = pB.next
        while b > -1 and pB:
            b -= 1
            pB = pB.next
        while p2end.next:
            p2end = p2end.next
        assert a == 1
        assert b == -1
        pA.next = list2
        p2end.next = pB
        return head

File: leet_LIST_1669_merge.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func mergeInBetween(list1 *ListNode, a int, b int, list2 *ListNode) *ListNode {
    head := list1
    p1a, p1b := head, head
    p2end := list2
    for a > 1 && p1a != nil {
        a--
        b--
        p1a = p1a.Next
        p1b = p1b.Next
    }
    for b > -1 && p1b != nil {
        b--
        p1b = p1b.Next
    }
    if a != 1 || b != -1 {
        panic(fmt.Sprintf("/a %d - /b %d", a, b))
    }
    for p2end.Next != nil {
        p2end = p2end.Next
    }
    p1a.Next, p2end.Next = list2, p1b
    return head
}
